### ISSUE TITLE
Windows Console builds seem to lack Accessibility Fixes

### ISSUE THREAD

**Issue by BlindGuyNW** [CleverRaven/Cataclysm-DDA#15882]
Per discussion [here](http://smf.cataclysmdda.com/index.php?topic=12309.30), the Windows console experimental builds of CDDA don't seem to include the accessibility fixes recently made to the Linux/curses version. Most critical is the cursor fix, so that the terminal cursor follows the @ symbol representing the PC. Hoping this and future accessibility fixes can be kept in sync with Windows, as the majority of blind gamers use it in one form or another.

**Comment by lucianposton** [comment]
The fix should have worked in both the linux and windows versions. It isn't immediately clear to me why the windows behavior would be different.

Anyone know a simple way to check the cursor position in a windows terminal w/o setting up a screen reader?

**Comment by BlindGuyNW** [comment]
It’s possible the problem is the screen reader console support. That would be incredibly annoying. I’m not sure how to confirm that though.

**Comment by lucianposton** [comment]
I wonder if wincurse.cpp's curses_drawwindow() simply fails to update the current cursor position (or causes the cursor to move to the wrong location)? If so, perhaps calling wmove() after wrefresh() in https://github.com/CleverRaven/Cataclysm-DDA/blob/661a938575210464beda73fb4f401d96050e651a/src/game.cpp#L5433-L5436 would be a quick way to fix the issue (or properly fixing curses_drawwindow).

I don't a way to test this on a windows machine though.

**Comment by BlindGuyNW** [comment]
I can definitely do some testing, though I don’t have a compiler set up for Windows myself. I’d be happy to help however I can.

**Comment by lucianposton** [comment]
@BlindGuyNW I got things cross-compiling and running in a windows virtual machine. I still don't know how to test where the cursor is though. How can I test that? Is there a certain screen reader and screen reader function I should use?

**Comment by lucianposton** [comment]
I did a bit of digging, and what I'm seeing is that the windows console build isn't a traditional windows console app. It uses a GUI window (instead of a console window). Consequently, the API to set the console cursor position does not work because the process doesn't even have a console attached. Maybe there is another API to set the cursor position in a non-console window, but I haven't yet found it on msdn.

I tried the JAWS screen reader on the experimental windows console build, and it didn't read anything (except for the window's title). I'm probably just using JAWS incorrectly, but does a screen reader work at all on the windows console build?

**Comment by BlindGuyNW** [comment]
That would explain quite a bit of what’s been puzzling us. 

Yes, Jaws will read the screen, but the key is to switch from the PC cursor to the Jaws cursor, which basically lets one have control over the mouse. To do so, press the minus key on the numeric keypad, or caps-lock-P (?) in the laptop keyboard layout. From there, using the keypad or arrow keys will let you explore. Insert-keypad minus is supposed to route the Jaws cursor/mouse to the PC cursor, but that doesn’t work in CDDA.

**Comment by lucianposton** [comment]
I see. Is that interaction, switching between JAWS and PC cursor to explore, even that accessible? As far as I can tell, in that mode, JAWS doesn't read screen updates (like the message log). So every move, you'd have to explore the screen. It seems like a traditional console is much more accessible.

At this point, the options to make the windows console build more accessible are

1. Keep the windows console build implemented as a GUI window. Update the mouse cursor to point to the character's position. Figure out why the JAWS cursor doesn't update the PC cursor, and fix that. I'm not even sure this is possible, and this seems to be much less accessible than if the app were a traditional windows console app.
2. Rewrite src/wincurse.cpp as a console, rather than a GUI. Much more work, but would be more accessible, similar to the linux console build.
3. Rewrite src/wincurse.cpp as a console using an ncurses port, e.g. https://github.com/Bill-Gray/PDCurses, licensing permitting. Similar to previous option, but it should save on dev work, compared to the previous option. However, it adds another dependency.

**Comment by BlindGuyNW** [comment]
No, unfortunately it isn’t that accessible. There are ways to script screen readers which would make the GUI more useable, but I don’t know much about that and in any case it’s highly SR-dependent.

I wonder what approach the Windows console ports of games like NetHack and Crawl Stone Soup take . I suspect it’s either 2 or 3. In any case, those ports are reasonably accessible.

All in all, this is more work than I imagined it would be. I wouldn’t blame CDDA if you decided to just focus on the Linux console as far as accessibility goes.

**Comment by h-rezaei** [comment]
Hi. I'm looking forward to play CDDA as a screen reader user. a while back some friends were talking about it and it sounded so exciting to me. Is it accessible to play with Linux?

**Comment by BlindGuyNW** [comment]
Hi,

Yes, the Linux version is pretty accessible if you take some time to learn the way the menus work and so forth. I use Speakup at the console and it plays very well. I haven’t messed with vehicles or advanced crafting too much, but it’s a lot nicer than the Windows version at this point.
Good luck. I need to get my Linux system back up and running so I can play.

**Comment by Night-Pryanik** [comment]
There isn't Windows curses version anymore, so closing.
